go/database/mpt/forest_test.go

```diff
@@ -2155,6 +2155,120 @@ func TestForest_VisitPathToStorage(t *testing.T) {
 	}
 }
 
+func TestForest_HasEmptyStorage_CastingDoesNotFail(t *testing.T) {
+	for _, variant := range variants {
+		for _, config := range allMptConfigs {
+			for forestConfigName, forestConfig := range forestConfigs {
+				t.Run(fmt.Sprintf("%s-%s-%s", variant.name, config.Name, forestConfigName), func(t *testing.T) {
+					directory := t.TempDir()
+					var err error
+					forest, err := variant.factory(directory, config, forestConfig)
+					if err != nil {
+						t.Fatalf("failed to open forest: %v", err)
+					}
+
+					root := NewNodeReference(ExtensionId(11))
+
+					// Test casting
+
+					isEmpty, err := forest.HasEmptyStorage(&root, common.Address{0x1})
+					if err != nil {
+						t.Fatalf("failed to ask whether account has empty storage: %v", err)
+					}
+					if !isEmpty {
+						t.Error("freshly created account has empty storage, but forest returned true")
+					}
+
+				})
+			}
+		}
+	}
+}
+
+func TestForest_HasEmptyStorage(t *testing.T) {
+	for _, variant := range variants {
+		for _, config := range allMptConfigs {
+			for forestConfigName, forestConfig := range forestConfigs {
+				t.Run(fmt.Sprintf("%s-%s-%s", variant.name, config.Name, forestConfigName), func(t *testing.T) {
+					directory := t.TempDir()
+					var err error
+					forest, err := variant.factory(directory, config, forestConfig)
+					if err != nil {
+						t.Fatalf("failed to open forest: %v", err)
+					}
+
+					root := NewNodeReference(EmptyId())
+
+					addresses := getTestAddresses(100)
+					for i, address := range addresses {
+						root, err = forest.SetAccountInfo(&root, address, AccountInfo{Balance: common.Balance{0x1}})
+						if err != nil {
+							t.Fatalf("cannot update account: %v", err)
+						}
+						// fill storage of half of the accounts
+						if i < 50 {
+							continue
+						}
+						root, err = forest.SetValue(&root, address, common.Key{byte(i)}, common.Value{byte(i)})
+						if err != nil {
+							t.Fatalf("cannot update storage: %v", err)
+						}
+					}
+
+					// Account with empty storage
+					isEmpty, err := forest.HasEmptyStorage(&root, addresses[0])
```

Comment:
I think it would be much stronger to test all accounts you created for them being empty or not. This cherrypicking of a few accounts is a bit awkward. 

You can repeat the loop and check that half is empty, half is not. During this loop you can empty more accounts using various strategies 
1/4 deleted by ClearStorage, 1/4 deleted by SetSlot, 1/4 deleted by SetAccountInfo, 1/4 remains non-empty
and after that re-check that expected accounts become empty.